app/controllers/locations_controller.rb

```diff
@@ -42,6 +42,45 @@ def update_ips
     end
   end
 
+  def bulk_upload
+    @upload_form = UploadForm.new
+  end
+
+  def upload_locations_csv
+    @upload_form = UploadForm.new(upload_params)
+    if @upload_form.invalid?
+      render :bulk_upload
+    else
+      @parent_organisation = current_organisation
+      BulkUpload::BulkUpload.add_location_data(@upload_form.data, @parent_organisation)
+      @parent_organisation.validate
+      # render
+    end
+  end
+
+  def confirm_upload
+    data = params.permit(csv: {}).fetch(:csv, {}).values.map(&:values)
+    raise "The uploaded file did not contain any locations." if data.empty?
+
+    @parent_organisation = current_organisation
+    BulkUpload::BulkUpload.add_location_data(data, @parent_organisation)
+    @parent_organisation.save!
+    redirect_to(ips_path, notice: "Successfully uploaded locations")
+  rescue ActiveRecord::RecordInvalid
+    redirect_to(ips_path, flash: { error: "Uploading data failed. Please try again." })
```

Comment:
Given the suggestion to the user here is to please try again, does this mean this record invalid exception relates to some situation that could be different if they try again with the same input?